packages/angular_devkit/build_angular/src/webpack/plugins/javascript-optimizer-plugin.ts

```diff
@@ -0,0 +1,178 @@
+   * all intermediate sourcemaps.
+   */
+  sourcemap: boolean;
+
+  /**
+   * The ECMAScript version that should be used when generated output code.
+   * The value must be lower case.
+   * Examples: `es5`, `es2015`, or `es2017`
+   */
+  target: string;
+
+  /**
+   * Enables the retention of identifier names and ensures that function and class names are
+   * present in the output code.
+   */
+  keepNames: boolean;
+
+  /**
+   * Enables the removal of all license comments from the output code.
+   */
+  removeLicenses: boolean;
+}
+
+/**
+ * A Webpack plugin that provides JavaScript optimization capabilities.
+ *
+ * The plugin uses both `esbuild` and `terser` to provide both fast and highly-optimized
+ * code output. `esbuild` is used as an initial pass to remove the majority of unused code
+ * as well as shorten identifiers. `terser` is then used as a secondary pass to apply
+ * optimizations not yet implemented by `esbuild`.
+ */
+export class JavaScriptOptimizerPlugin {
+  constructor(public options: Partial<JavaScriptOptimizerOptions> = {}) {}
+
+  apply(compiler: import('webpack').Compiler) {
+    const { OriginalSource, SourceMapSource } = compiler.webpack.sources;
+
+    compiler.hooks.compilation.tap(PLUGIN_NAME, (compilation) => {
+      compilation.hooks.processAssets.tapPromise(
+        {
+          name: PLUGIN_NAME,
+          stage: compiler.webpack.Compilation.PROCESS_ASSETS_STAGE_OPTIMIZE_SIZE,
+        },
+        async (compilationAssets) => {
+          const scriptsToOptimize = [];
+
+          // Analyze the compilation assets for scripts that require optimization
+          for (const assetName of Object.keys(compilationAssets)) {
+            if (assetName.endsWith('.js')) {
+              const scriptAsset = compilation.getAsset(assetName);
+              if (scriptAsset && !scriptAsset.info.minimized) {
+                const { source, map } = scriptAsset.source.sourceAndMap();
+                scriptsToOptimize.push({
+                  name: scriptAsset.name,
+                  code: typeof source === 'string' ? source : source.toString(),
+                  map,
+                });
+              }
+            }
+          }
+
+          if (scriptsToOptimize.length === 0) {
+            return;
+          }
+
+          // Ensure all replacement values are strings which is the expected type for esbuild
+          let define: Record<string, string> | undefined;
+          if (this.options.define) {
+            define = {};
+            for (const [key, value] of Object.entries(this.options.define)) {
+              define[key] = String(value);
+            }
+          }
+
+          // Setup the options used by all worker tasks
+          const optimizeOptions = {
+            sourcemap: this.options.sourcemap,
+            define,
+            keepNames: this.options.keepNames,
+            target: this.options.target ? Number(this.options.target.slice(2)) : 2017,
```

Comment:
`Number(this.options.target.slice(2))` would be `NaN` when target is `ESNext` or `Latest`.

Reply:
Adjusted the logic to prevent those cases from happening.